msprime/simulations.py

```diff
@@ -1365,6 +1368,139 @@ def print_history(self, output=sys.stdout):
             self._print_populations(epoch, output)
             print(file=output)
 
+    def population_size_trajectory(self, end, num_steps=10):
+        """
+        This function computes the defined effective population size
+        for each sub-population, for any demographic model.
+
+        `end` specifies at which generation to stop
+        collecting effective population sizes among subpopulations.
+        The sampling will always begin at generation 0
+
+        `num_steps` parameter will determine how many points from
+        0 to `end` to sample.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a
+        num_steps-by-num_pops ndarray that will contain the effective
+        population sizes for each subpopulation individually, for each step,
+        as defined by the model which inherits from this class.
+        """
+
+        num_pops = self.num_populations
+        N_t = np.zeros([num_steps, num_pops])
+        steps = np.linspace(0, end, num_steps)
+        for j, t in enumerate(steps):
+            N = self.pop_size_and_migration_at_t(t)["population_sizes"]
+            N_t[j] = N
+        return steps, N_t
+
+    def coalescence_rate_trajectory(self, end, num_samples,
+                                    num_steps=10, min_pop_size=1):
+        """
+        This function will calculate the ground truth
```

Comment:
Edits:
```
        This function will calculate the mean coalescence rates between the lineages of the 
        sample configuration provided in `num_samples`, at a regularly spaced set of times,
        in this demographic model. The coalescence rate at time t in the past is the average rate of coalescence of 
        as-yet-uncoalesed lineages, computed as follows: let p(t) be the probability that the
        lineages of a randomly chosen pair of samples has not yet coalesced by time t, 
        let p(z,t) be the probability that the lineages of a randomly chosen pair of samples 
        has not yet coalesced by time t *and* are both in population z, and let N(z,t) be the
        diploid effective population size of population z at time t. Then the mean coalescence
        rate at time t is r(t) = (sum_z p(z,t) / (2 * N(z,t) ) / p(t) .

        The computation is done approximating population size trajetories with piecewise constant
        trajectories, in steps of size dt = end / num_steps. For this to be accurate, dt must be small 
        enough so that (a) short epochs (e.g., bottlenecks) are not missed, and (b) populations do not
        change in size too much over that time, if they are growing or shrinking. If in doubt, check that
        results are simliar after doubling `num_steps`.

        `num_samples` is a list of the same length as the number of populations, so that `num_samples[j]` is the number of sampled chromosomes in subpopulation `j`.
        `num_steps` determines at how many time points the coalescence rate is computed.
        `min_pop_size` is the smallest allowed population size during computation of coalescent rates (i.e.,
             coalescence rates are actually 1 / 2 * max(min_pop_size, N(z,t))). Spurious very small population sizes
             can occurIn models where populations grow exponentially but are unused before some time in the past,
             and lead to floating point error. This should be set to a value smaller than the smallest desired population size in the model.

        Returns a tuple where the first value
        is an array of time points, the second is an array whose jth element is the coalescence rate at the jth time
        point (denoted r(t[j]) above), and the third is the probablility that a randomly pair of lineages
        has not yet coalesced (denoted p(t[j]) above).
```

Reply:
super cool!!!

quick comment. the phrase "...in models where populations grow exponentially but are unused..."-- do you mean unused in the sense that neither of the lineages has been able to migrate into the focal population?

Reply:
I mean unused like there was a mass migration event that moved everyone out of that population already, and if you drew the population diagram it wouldn't even be there. We should probably put in an example. Like if in forwards time, population 0 split into populations 0 and 1, a thousand years ago; then since msprime always has the same number of pops around, older than 1000 years ago pop 1 would be there, but be 'unused'.